README.md
# zk-shell [![Build Status] (https://travis-ci.org/rgs1/zk_shell.svg?branch=master)](https://travis-ci.org/rgs1/zk_shell) [![Coverage Status](https://coveralls.io/repos/rgs1/zk_shell/badge.png?branch=master)](https://coveralls.io/r/rgs1/zk_shell?branch=master) [![PyPI version](https://badge.fury.io/py/zk_shell.svg)](http://badge.fury.io/py/zk_shell)

**Table of Contents**

- [tl;dr](#tldr)
- [Installing](#installing)
- [Usage](#usage)
- [Dependencies](#dependencies)

### tl;dr ###

A powerful & scriptable shell for [Apache ZooKeeper](http://zookeeper.apache.org/ "ZooKeeper")

### Installing ###

From PyPI:

```
$ pip install zk-shell
```

Or running from the source:

```
# Kazoo is needed
$ pip install kazoo

$ git clone https://github.com/rgs1/zk_shell.git
$ cd zk_shell
$ export ZKSHELL_SRC=1; bin/zk-shell
Welcome to zk-shell (0.99.04)
(DISCONNECTED) />
```

You can also build a self-contained PEX file:

```
$ pip install pex

$ pex -v -e zk_shell.cli -o zk-shell.pex .

```

More info about PEX [here](https://pex.readthedocs.org "PEX").

### Usage ###

```
$ zk-shell localhost:2181
(CONNECTED) /> ls
zookeeper
(CONNECTED) /> create foo 'bar'
(CONNECTED) /> get foo
bar
(CONNECTED) /> cd foo
(CONNECTED) /foo> create ish 'barish'
(CONNECTED) /foo> cd ..
(CONNECTED) /> ls foo
ish
(CONNECTED) /> create temp- 'temp' true true
(CONNECTED) /> ls
zookeeper foo temp-0000000001
(CONNECTED) /> rmr foo
(CONNECTED) />
(CONNECTED) /> tree
.
├── zookeeper
│   ├── config
│   ├── quota
```

Line editing and command history is supported via readline (if readline is available). There's
also autocomplete for most commands and their parameters.

Individual files can be copied between the local filesystem and ZooKeeper. Recursively copying
from the filesystem to ZooKeeper is supported as well, but not the other way around since znodes
can have content and children.

```
(CONNECTED) /> cp file:///etc/passwd zk://localhost:2181/passwd
(CONNECTED) /> get passwd
(...)
unbound:x:992:991:Unbound DNS resolver:/etc/unbound:/sbin/nologin
haldaemon:x:68:68:HAL daemon:/:/sbin/nologin
```

Copying between one ZooKeeper cluster to another is supported, too:

```
(CONNECTED) /> cp zk://localhost:2181/passwd zk://othercluster:2183/mypasswd
```

Copying between a ZooKeeper cluster and JSON files is supported as well:

```
(CONNECTED) /> cp zk://localhost:2181/something json://!tmp!backup.json/ true true
```

Mirroring paths to between clusters or JSON files is also supported. Mirroring replaces the destination
path with the content and structure of the source path.

```
(CONNECTED) /> create /source/znode1/znode11 'Hello' false false true
(CONNECTED) /> create /source/znode2 'Hello' false false true
(CONNECTED) /> create /target/znode1/znode12 'Hello' false false true
(CONNECTED) /> create /target/znode3 'Hello' false false true
(CONNECTED) /> tree
.
├── target
│   ├── znode3
│   ├── znode1
│   │   ├── znode12
├── source
│   ├── znode2
│   ├── znode1
│   │   ├── znode11
├── zookeeper
│   ├── config
│   ├── quota
(CONNECTED) /> mirror /source /target
Are you sure you want to replace /target with /source? [y/n]:
y
Mirroring took 0.04 secs
(CONNECTED) /> tree
.
├── target
│   ├── znode2
│   ├── znode1
│   │   ├── znode11
├── source
│   ├── znode2
│   ├── znode1
│   │   ├── znode11
├── zookeeper
│   ├── config
│   ├── quota
(CONNECTED) /> create /target/znode4 'Hello' false false true
(CONNECTED) /> mirror /source /target false false true
Mirroring took 0.03 secs
(CONNECTED) />
```

Debugging watches can be done with the watch command. It allows monitoring all the child watches
that, recursively, fire under <path>:

```
(CONNECTED) /> watch start /
(CONNECTED) /> create /foo 'test'
(CONNECTED) /> create /bar/foo 'test'
(CONNECTED) /> rm /bar/foo
(CONNECTED) /> watch stats /

Watches Stats

/foo: 1
/bar: 2
/: 1
(CONNECTED) /> watch stop /
```

Searching for paths or znodes which match a given text can be done via find:

```
(CONNECTED) /> find / foo
/foo2
/fooish/wayland
/fooish/xorg
/copy/foo
```

Or a case-insensitive match using ifind:

```
(CONNECTED) /> ifind / foo
/foo2
/FOOish/wayland
/fooish/xorg
/copy/Foo
```

Grepping for content in znodes can be done via grep:

```
(CONNECTED) /> grep / unbound true
/passwd: unbound:x:992:991:Unbound DNS resolver:/etc/unbound:/sbin/nologin
/copy/passwd: unbound:x:992:991:Unbound DNS resolver:/etc/unbound:/sbin/nologin
```

Or via igrep for a case-insensitive version.

Non-interactive mode can be used passing commands via --run-once:

```
$ zk-shell --run-once "create /foo 'bar'" localhost
$ zk-shell --run-once "get /foo" localhost
bar
```

Or piping commands through stdin:

```
$ echo "get /foo" | zk-shell --run-from-stdin localhost
bar
```

It's also possible to connect using an SSH tunnel, by specifying a host to use:

```
$ zk-shell --tunnel ssh-host zk-host
```

### Dependencies ###

* Python 2.7, 3.3 or 3.4
* Kazoo >= 2.2

### Testing and Development ###

Please see [CONTRIBUTING.md](CONTRIBUTING.md).


CHANGES.md
ChangeLog
=========

1.1.1 (2015-XX-XX)
--------------------

Bug Handling
************
- fix doc error in `sleep`'s documentation
- fix NameError in xclient when dns lookups fail

Features
********
- add `pretty_date` option for `exists` command
- print zxids in `exists` as hex
- all boolean parameters now support a label, i.e.:

  ```
  (CONNECTED) /> ls / watch=true
  ```
- new `time` command to measure execution (time) of the given commands
- the `create` command now supports async mode (async=true)
- print last_zxid in `session_info` as hex
- the `session_info` commands now has an optional [match] parameter
- new command `echo` to print formatted strings with extrapolated commands

1.1.0 (2015-06-17)
--------------------

Bug Handling
************
- handle APIError (i.e.: ZooKeeper internal error)

Features
********
- add --version
- add `stat` alias for `exists` command
- add reconfig command (as offered by ZOOKEEPER-107)

1.0.08 (2015-06-05)
--------------------

Bug Handling
************
-

Features
********
- allow connecting via an ssh tunnel (--tunnel)

1.0.07 (2015-06-03)
--------------------

Bug Handling
************
- issue with tree command output (issue #28)
- intermittent issue with child_count (issue #30)

Features
********
- sleep: allows sleeping (useful with loop)

1.0.06 (2015-05-06)
--------------------

Bug Handling
************
- don't allow running edit as root
- default to /usr/bin/vi for edit
- check that the provided editor is executable
- don't trust editor commands that are setuid/setgid
- treat None as "" when using the `edit` command

Features
********
- add `man` alias for `help` command
- improve docstrings & use man pages style

1.0.05 (2015-04-09)
--------------------

Bug Handling
************
-

Features
********
- edit: allows inline editing of a znode

1.0.04 (2015-04-02)
--------------------

Bug Handling
************
- fix bad variable reference when handling bad JSON keys
- ls: always sort znodes

Features
********
- json_dupes_for_keys: finds duplicated znodes for the given keys
- pipe: pipe commands (though more like xargs -n1)

1.0.03 (2015-02-24)
--------------------

Bug Handling
************
- fix race condition in chkzk

Features
********
- add conf command to configure runtime variables
- chkzk: show states

1.0.02 (2015-02-12)
--------------------

Bug Handling
************
- handle bad (non-closed) quotations in commented commands
- improve `watch`'s documentation

Features
********
- show help when a command is wrong or missing params
- add chkzk to check if a cluster is in a consistent state

1.0.01 (2014-12-31)
--------------------

Bug Handling
************
- fix rm & rmr from relative paths (issue #11)

Features
********
-

1.0.0 (2014-12-24)
--------------------

Bug Handling
************
- fix async cp
- fix off-by-one for summary of /
- allow creating sequential znodes when the base path exists
- don't crash grep when znodes have no bytes (None)

Features
********
- better coverage
- rm & rmr now take multiple paths
- transactions are now supported

0.99.05 (2014-12-08)
--------------------

Bug Handling
************

- to allow a 3rd param in set_acls, acls must be quoted now
- don't crash in add_auth when the scheme is unknown (AuthFailedError)
- don't crash in cp when the scheme is unknown (AuthFailedError)
- handle IPv6 addresses within cp commands (reported by @fsparv)

Features
********

- the acls params in set_acls now need to be quoted
- set_acls now supports recursive mode via a 3rd optional param
- TravisCI is now enabled so tests should always run
- suggest possible commands when the command is unknown

0.99.04 (2014-11-25)
--------------------

Bug Handling
************

- Examples for mntr, cons & dump
- Fix autocomplete when the path isn't the 1st param
- Fix path completion when outside of /

Features
********

- New shortcuts for cd


zk_shell/copy.py
""" helpers to move files/dirs to and from ZK and also among ZK clusters """

from __future__ import print_function

from base64 import b64decode, b64encode
from collections import defaultdict
import json
import os
import re
import time
import shutil

try:
    from urlparse import urlparse
except ImportError:
    # Python 3.3?
    from urllib.parse import urlparse

from kazoo.client import KazooClient
from kazoo.exceptions import (
    NoAuthError,
    NodeExistsError,
    NoNodeError,
    NoChildrenForEphemeralsError,
    ZookeeperError,
)

from .acl import ACLReader
from .statmap import StatMap
from .util import join, Netloc, to_bytes


DEFAULT_ZK_PORT = 2181


def zk_client(host, scheme, credential):
    """ returns a connected (and possibly authenticated) ZK client """

    if not re.match(r".*:\d+$", host):
        host = "%s:%d" % (host, DEFAULT_ZK_PORT)

    client = KazooClient(hosts=host)
    client.start()

    if scheme != "":
        client.add_auth(scheme, credential)

    return client


class CopyError(Exception):
    """ base exception for Copy errors """

    def __init__(self, message, early_error=False):
        super(CopyError, self).__init__(message)
        self._early_error = early_error

    @property
    def is_early_error(self):
        return self._early_error


class AuthError(CopyError):
    """ authentication exception for Copy """
    def __init__(self, operation, path):
        super(AuthError, self).__init__(
            "Permission denied: Could not %s znode %s." % (operation, path))


class PathValue(object):
    def __init__(self, value, acl=None):
        self._value = value
        self._acl = acl if acl else []

    @property
    def value(self):
        return self._value

    @property
    def value_as_bytes(self):
        return to_bytes(self.value)

    @property
    def acl(self):
        return self._acl

    @property
    def acl_as_dict(self):
        return self._acl


class ProxyType(type):
    TYPES = {}
    SCHEME = ""

    def __new__(mcs, clsname, bases, dct):
        obj = super(ProxyType, mcs).__new__(mcs, clsname, bases, dct)
        if obj.SCHEME in mcs.TYPES:
            raise ValueError("Duplicate scheme handler: %s" % obj.SCHEME)

        if obj.SCHEME != "":
            mcs.TYPES[obj.SCHEME] = obj
        return obj


class Proxy(ProxyType("ProxyBase", (object,), {})):
    SCHEME = ""

    def __init__(self, parse_result, exists, async, verbose):
        self.parse_result = parse_result
        self.netloc = Netloc.from_string(parse_result.netloc)
        self.exists = exists
        self.async = async
        self.verbose = verbose

    @property
    def scheme(self):
        return self.parse_result.scheme

    @property
    def url(self):
        return self.parse_result.geturl()

    @property
    def path(self):
        path = self.parse_result.path
        if path == "":
            return "/"
        return "/" if path == "/" else path.rstrip("/")

    @property
    def host(self):
        return self.netloc.host

    @property
    def auth_scheme(self):
        return self.netloc.scheme

    @property
    def auth_credential(self):
        return self.netloc.credential

    def set_url(self, string):
        """ useful for recycling a stateful proxy """
        self.parse_result = Proxy.parse(string)

    @classmethod
    def from_string(cls, string, exists=False, async=False, verbose=False):
        """
        if exists is bool, then check it either exists or it doesn't.
        if exists is None, we don't care.
        """
        result = cls.parse(string)

        if result.scheme not in cls.TYPES:
            raise CopyError("Invalid scheme: %s" % (result.scheme))

        return cls.TYPES[result.scheme](result, exists, async, verbose)

    @classmethod
    def parse(cls, url_string):
        return urlparse(url_string)

    def __enter__(self):
        pass

    def __exit__(self, etype, value, traceback):
        pass

    def check_path(self):
        raise NotImplementedError("check_path must be implemented")

    def read_path(self):
        raise NotImplementedError("read_path must be implemented")

    def write_path(self, path_value):
        raise NotImplementedError("write_path must be implemented")

    def children_of(self):
        raise NotImplementedError("children_of must be implemented")

    def delete_path_recursively(self):
        raise NotImplementedError("delete_path must be implemented")

    def copy(self, dst, recursive, max_items, mirror):
        opname = "Copy" if not mirror else "Mirror"

        # basic sanity check
        if mirror and self.scheme == "zk" and dst.scheme == "file":
            raise CopyError("Mirror from zk to fs isn't supported", True)

        if recursive and self.scheme == "zk" and dst.scheme == "file":
            raise CopyError("Recursive %s from zk to fs isn't supported" %
                            opname.lower(), True)

        if mirror and not recursive:
            raise CopyError("Mirroring must be recursive", True)

        start = time.time()

        src_url = self.url
        dst_url = dst.url

        with self:
            with dst:
                if mirror:
                    dst_children = set(c for c in dst.children_of())

                self.do_copy(dst, opname)

                if recursive:
                    for i, child in enumerate(self.children_of()):
                        if mirror and child in dst_children:
                            dst_children.remove(child)
                        if max_items > 0 and i == max_items:
                            break
                        self.set_url(join(src_url, child))
                        dst.set_url(join(dst_url, child))
                        self.do_copy(dst, opname)

                        # reset to base urls
                        self.set_url(src_url)
                        dst.set_url(dst_url)

                if mirror:
                    for child in dst_children:
                        dst.set_url(join(dst_url, child))
                        dst.delete_path_recursively()

        end = time.time()

        print("%sing took %.2f secs" % (opname, round(end - start, 2)))

    def do_copy(self, dst, opname):
        if self.verbose:
            if self.async:
                print("%sing (asynchronously) from %s to %s" % (opname, self.url, dst.url))
            else:
                print("%sing from %s to %s" % (opname, self.url, dst.url))

        dst.write_path(self.read_path())


class ZKProxy(Proxy):
    """ read/write ZooKeeper paths """

    SCHEME = "zk"

    class ZKPathValue(PathValue):
        """ handle ZK specific meta attribs (i.e.: acls) """
        def __init__(self, value, acl=None):
            PathValue.__init__(self, value)
            self._acl = acl

        @property
        def acl(self):
            return self._acl

        @property
        def acl_as_dict(self):
            acls = self.acl if self.acl else []
            return [ACLReader.to_dict(a) for a in acls]

    def __init__(self, parse_result, exists, async, verbose):
        super(ZKProxy, self).__init__(parse_result, exists, async, verbose)
        self.client = None
        self.need_client = True  # whether we build a client or one is provided

    def connect(self):
        if self.need_client:
            self.client = zk_client(self.host, self.auth_scheme, self.auth_credential)

    def disconnect(self):
        if self.need_client:
            if self.client:
                self.client.stop()

    def __enter__(self):
        self.connect()

        if self.exists is not None:
            self.check_path()

    def __exit__(self, etype, value, traceback):
        self.disconnect()

    def check_path(self):
        try:
            retval = True if self.client.exists(self.path) else False
        except NoAuthError:
            raise AuthError("read", self.path)

        if retval is not self.exists:
            if self.exists:
                error = "znode %s in %s doesn't exist" % \
                    (self.path, self.host)
            else:
                error = "znode %s in %s exists" % (self.path, self.host)
            raise CopyError(error)

    def read_path(self):
        try:
            # TODO: propose a new ZK opcode (GetWithACLs) so we can do this in 1 rt
            value = self.get_value(self.path)
            acl, _ = self.client.get_acls(self.path)
            return self.ZKPathValue(value, acl)
        except NoAuthError:
            raise AuthError("read", self.path)

    def write_path(self, path_value):
        if isinstance(path_value, self.ZKPathValue):
            acl = path_value.acl
        else:
            acl = [ACLReader.from_dict(a) for a in path_value.acl]

        if self.client.exists(self.path):
            try:
                value = self.get_value(self.path)
                if path_value.value != value:
                    self.client.set(self.path, path_value.value)
            except NoAuthError:
                raise AuthError("write", self.path)
        else:
            try:
                # Kazoo's create() doesn't handle acl=[] correctly
                # See: https://github.com/python-zk/kazoo/pull/164
                acl = acl or None
                self.client.create(self.path, path_value.value, acl=acl, makepath=True)
            except NoAuthError:
                raise AuthError("create", self.path)
            except NodeExistsError:
                raise CopyError("Node %s exists" % (self.path))
            except NoNodeError:
                raise CopyError("Parent node for %s is missing" % (self.path))
            except NoChildrenForEphemeralsError:
                raise CopyError("Ephemeral znodes can't have children")
            except ZookeeperError:
                raise CopyError("ZooKeeper server error")

    def get_value(self, path):
        try:
            if hasattr(self.client, 'get_bytes'):
                v, _ = self.client.get_bytes(path)
            else:
                v, _ = self.client.get(path)
        except NoAuthError:
            raise AuthError("read", path)

        return v

    def delete_path_recursively(self):
        try:
            self.client.delete(self.path, recursive=True)
        except NoNodeError:
            pass
        except NoAuthError:
            raise AuthError("delete", self.path)
        except ZookeeperError:
            raise CopyError("Zookeeper server error")

    def children_of(self):
        if self.async:
            offs = 1 if self.path == "/" else len(self.path) + 1
            for path, stat in StatMap(self.client, self.path, recursive=True).get():
                if stat.ephemeralOwner == 0:
                    yield path[offs:]
        else:
            for path in self.zk_walk(self.path, None):
                yield path

    def zk_walk(self, root_path, branch_path):
        """
        skip ephemeral znodes since there's no point in copying those
        """
        full_path = join(root_path, branch_path) if branch_path else root_path

        try:
            children = self.client.get_children(full_path)
        except NoNodeError:
            children = set()
        except NoAuthError:
            raise AuthError("read children", full_path)

        for child in children:
            child_path = join(branch_path, child) if branch_path else child

            try:
                stat = self.client.exists(join(root_path, child_path))
            except NoAuthError:
                raise AuthError("read", child)

            if stat is None or stat.ephemeralOwner != 0:
                continue
            yield child_path
            for new_path in self.zk_walk(root_path, child_path):
                yield new_path

class FileProxy(Proxy):
    SCHEME = "file"

    def __init__(self, parse_result, exists, async, verbose):
        super(FileProxy, self).__init__(parse_result, exists, async, verbose)

        if exists is not None:
            self.check_path()

    def check_path(self):
        if os.path.exists(self.path) is not self.exists:
            error = "Path %s " % (self.path)
            error += "doesn't exist" if self.exists else "exists"
            raise CopyError(error)

    def read_path(self):
        if os.path.isfile(self.path):
            with open(self.path, "r") as fph:
                return PathValue("".join(fph.readlines()))
        elif os.path.isdir(self.path):
            return PathValue("")

        raise CopyError("%s is of unknown file type" % (self.path))

    def write_path(self, path_value):
        """ this will overwrite dst path - be careful """
        parent_dir = os.path.dirname(self.path)
        try:
            os.makedirs(parent_dir)
        except OSError:
            pass
        with open(self.path, "w") as fph:
            fph.write(path_value.value)

    def children_of(self):
        root_path = self.path[0:-1] if self.path.endswith("/") else self.path
        for path, _, files in os.walk(root_path):
            path = path.replace(root_path, "")
            if path.startswith("/"):
                path = path[1:]
            if path != "":
                yield path
            for filename in files:
                yield join(path, filename) if path != "" else filename

    def delete_path_recursively(self):
        shutil.rmtree(self.path, True)


class JSONProxy(Proxy):
    """ read/write from JSON files discovered via:

          json://!some!path!backup.json/some/path

        the serialized version looks like this:

        .. code-block:: python

         {
          '/some/path': {
             'content': 'blob',
             'acls': []},
          '/some/other/path': {
             'content': 'other-blob',
             'acls': []},
         }

        For simplicity, a flat dictionary is used as opposed as
        using a tree like format with children accessible from
        their parent.
    """

    def __init__(self, *args, **kwargs):
        super(JSONProxy, self).__init__(*args, **kwargs)
        self._dirty = None
        self._tree = None

    SCHEME = "json"

    def __enter__(self):
        self._dirty = False  # tracks writes

        self._tree = defaultdict(dict)
        if os.path.exists(self.host):
            with open(self.host, "r") as fph:
                try:
                    ondisc_tree = json.load(fph)
                    self._tree.update(ondisc_tree)
                except ValueError:
                    pass

        if self.exists is not None:
            self.check_path()

    def __exit__(self, etype, value, traceback):
        if not self._dirty:
            return

        with open(self.host, "w") as fph:
            json.dump(self._tree, fph, indent=4)

    @property
    def host(self):
        return super(JSONProxy, self).host.replace("!", "/")

    def check_path(self):
        if (self.path in self._tree) != self.exists:
            error = "Path %s " % (self.path)
            error += "doesn't exist" if self.exists else "exists"
            raise CopyError(error)

    def read_path(self):
        value = self._tree[self.path]["content"]
        if value is not None:
            try:
                value = b64decode(value)
            except:
                print("Failed to b64decode %s" % self.path)

        acl = self._tree[self.path].get("acls", [])
        return PathValue(value, acl)

    def write_path(self, path_value):
        content = path_value.value_as_bytes
        if content is not None:
            try:
                content = b64encode(content).decode(encoding="utf-8")
            except:
                print("Failed to b64encode %s" % self.path)

        self._tree[self.path]["content"] = content
        self._tree[self.path]["acls"] = path_value.acl_as_dict
        self._dirty = True

    def children_of(self):
        offs = 1 if self.path == "/" else len(self.path) + 1
        good = lambda k: k != self.path and k.startswith(self.path)
        for child in self._tree.keys():
            if good(child):
                yield child[offs:]

    def delete_path_recursively(self):
        if self.path in self._tree:
            # build a set from the iterable so we don't change the dictionary during iteration
            for c in set(self.children_of()):
                self._tree.pop(join(self.path, c))
            self._tree.pop(self.path)
